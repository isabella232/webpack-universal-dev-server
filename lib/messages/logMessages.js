// @flow

import { EOL } from "os";
import chalk from "chalk";

const separator = "---------------------------------------------------------------------------------------";
const finished = (
    what: string
): string =>
    chalk.green(`${ what } finished`);
const formatDuration = (
    duration: number
): string =>
    (duration / 1000).toFixed(1);

export const buildStarted = (
    what: string
): string =>
    `${ chalk.yellow(what) } started...`;

export const buildSuccess = (
    what: string,
    duration: number
): string =>
    `${ finished(what) } after ${ formatDuration(duration) }s`;

export const buildError = (
    what: string,
    error: string
): string =>
    [
        "",
        separator,
        chalk.red(`${ what } error`),
        separator,
        "",
        error,
        "",
        separator,
    ].join(EOL);
