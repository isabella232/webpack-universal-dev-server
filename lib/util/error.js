// @flow

import chalk from "chalk";

type AttemptWhatResult = *;

type AttemptOnErrorResult = *;

type AttemptArgs = {
    +what: () => AttemptWhatResult,
    +onError: (e: Error) => AttemptOnErrorResult
};

export class WUDSError extends Error {
    // the super() call cannot be placed inside the void operator, at least with babel
    constructor(message: string): void { // eslint-disable-line max-statements
        super(chalk.red.bold(message));

        return void (
            this.name = "WUDSError"
        );
    }
}

// TODO: Remove this nonsense
export const throwError = (e: Error) => {
    throw e;
};

// TODO: Remove this nonsense
// this is not expressable with just one statement
export const attempt = ({ // eslint-disable-line max-statements
    what,
    onError,
}: AttemptArgs): (AttemptWhatResult | AttemptOnErrorResult) => {
    try {
        return what();
    } catch (e) {
        return onError(e);
    }
};
