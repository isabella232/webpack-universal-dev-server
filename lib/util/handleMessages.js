// @flow

import type { Process } from "flow";
import type { Message, MessageHandlers } from "../types";

type ProcessMessage = Message | any;

export default function handleMessages(process: Process, handlers: MessageHandlers) {
    process.on("message", (message: ?ProcessMessage) => {
        if (message && typeof message.type === "string") {
            const handler = handlers.get(message.type);

            handler && handler(message);
        }
    });
}
