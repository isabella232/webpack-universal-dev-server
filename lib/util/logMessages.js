// @flow

import { EOL } from "os";
import chalk from "chalk";

const separator = "---------------------------------------------------------------------------------------";

export function buildStarted(what: string): string {
    return `${ chalk.yellow(what) } started...`;
}

export function buildSuccess(what: string, duration: number): string {
    const green = chalk.green(`${ what } finished`);

    return `${ green } after ${ (duration / 1000).toFixed(1) }s`;
}

export function buildError(what: string, error: string): string {
    return [
        "",
        separator,
        chalk.red(`${ what } error`),
        separator,
        "",
        error,
        "",
        separator
    ].join(EOL);
}
