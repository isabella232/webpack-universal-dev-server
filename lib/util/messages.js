// @flow

const NAMESPACE = __filename;

// WDS = Webpack Dev Server (watches, compiles and serves the client bundle)
// WWP = Webpack Watch Process (watches and compiles the server bundle)
export const TYPE_WDS_PAUSE = `${ NAMESPACE }/WDS/PAUSE`;
export const TYPE_WDS_RESUME = `${ NAMESPACE }/WDS/RESUME`;
export const TYPE_WWP_COMPILATION = `${ NAMESPACE }/WWP/COMPILATION`;
export const TYPE_WWP_DONE = `${ NAMESPACE }/WWP/DONE`;

export type Message = {
    type: string;
};

export function wdsPause(): Message {
    return {
        type: TYPE_WDS_PAUSE
    };
}

export function wdsResume(): Message {
    return {
        type: TYPE_WDS_RESUME
    };
}

export function wwpCompilation(): Message {
    return {
        type: TYPE_WWP_COMPILATION
    };
}

export function wwpDone(): Message {
    return {
        type: TYPE_WWP_DONE
    };
}
