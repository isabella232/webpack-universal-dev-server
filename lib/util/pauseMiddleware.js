// @flow

import type { Message } from "./messages";
import { TYPE_PAUSE, TYPE_RESUME } from "./messages";
import Deferred from "./Deferred";

type Request = {};
type Response = {};

const handlers: Map<string, Function> = new Map();
let deferred: Deferred | null = null;

function onMessage(message: Message) {
    const handler = message && message.type && handlers.get(message.type);

    handler && handler(message);
}

export default function pauseMiddleware(req: Request, res: Response, next: Function) {
    if (!deferred) {
        next();
        return;
    }
    deferred.promise.then(next, next);
}

handlers.set(TYPE_PAUSE, () => {
    if (!deferred) {
        deferred = new Deferred();
    }
});
handlers.set(TYPE_RESUME, () => {
    deferred && deferred.resolve();
    deferred = null;
});

process.on("message", onMessage);
