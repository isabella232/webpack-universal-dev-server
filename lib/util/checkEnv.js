// @flow

import { WUDSError } from "./error";

type ProcessEnv = { [key: string]: ?string };
type RequiredKeys = Array<string>;

export default (
    env: ProcessEnv,
    requiredKeys: RequiredKeys
): void => {
    void requiredKeys.forEach(key => {
        if (key in env === false) {
            throw new WUDSError(`The required env variable '${ key }' is not defined on process.env: Please specify it when creating the child_process`);
        }
    });
};
