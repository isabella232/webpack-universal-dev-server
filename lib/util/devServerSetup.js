// @flow

import type { WebpackDevServerSetup, ExpressApp } from "../types";
import pauseMiddleware from "../util/pauseMiddleware";

export default function devServerSetup(originalSetup?: WebpackDevServerSetup): WebpackDevServerSetup {
    return (...args) => {
        const app: ExpressApp = args[0];

        app.use(pauseMiddleware);

        return originalSetup ?
            // disabling some linting rules because we just want to preserve the original behavior as close as possible
            originalSetup.apply(this, args) : // eslint-disable-line no-invalid-this
            undefined; // eslint-disable-line no-undefined
    };
}
