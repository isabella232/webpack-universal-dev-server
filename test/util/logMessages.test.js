import { EOL } from "os";
import test from "ava";
import { buildStarted, buildSuccess, buildError } from "../../lib/util/logMessages";

// Colors cannot be tested in a good way because most CI systems don't use colors.
// We split the message by EOL to make these snapshot tests OS independent.
function check(message) {
    return t => {
        t.snapshot(message.split(EOL));
    };
}

test(
    "should log the expected buildStarted",
    check(buildStarted("webpack client build"))
);

test(
    "should log the expected buildSuccess message with durations above 1 second",
    check(buildSuccess("webpack client build", 1300))
);

test(
    "should log the expected buildSuccess message with durations below 1 second",
    check(buildSuccess("webpack client build", 300))
);

test(
    "should log the expected buildError message",
    check(buildError("webpack client build", "Module build failed"))
);
