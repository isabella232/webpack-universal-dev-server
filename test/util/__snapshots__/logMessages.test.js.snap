exports[`should log the expected buildError message 1`] = `"["","---------------------------------------------------------------------------------------","webpack client build error","---------------------------------------------------------------------------------------","","Module build failed","","---------------------------------------------------------------------------------------"]"`;

exports[`should log the expected buildStarted 1`] = `"[\"webpack client build started...\"]"`;

exports[`should log the expected buildSuccess message with durations above 1 second 1`] = `"[\"webpack client build finished after 1.3s\"]"`;

exports[`should log the expected buildSuccess message with durations below 1 second 1`] = `"[\"webpack client build finished after 0.3s\"]"`;
